File: libs/jgdsl/src/gdsl/asm/operand/Scale.java
package gdsl.asm.operand;

import gdsl.asm.Visitor;

public class Scale extends Operand {
  private long factor;
  
  public long getFactor() {
    return factor;
  }
  
  private Operand rhs;
  
  public Operand getRhs() {
    return rhs;
  }

  /**
   * @param lhs
   * @param rhs
   */
  public Scale (long factor, Operand rhs) {
    super();
    this.factor = factor;
    this.rhs = rhs;
  }

  @Override public String toString () {
    return factor + "*(" + rhs + ")";
  }

  @Override public void accept (Visitor v) {
    v.visit(this);
  }
}

File: libs/cppgdsl/src/assembly/operand/scale.cpp
#include "cppgdsl/assembly/operand/scale.h"
#include "cppgdsl/assembly/operand/operand_visitor.h"

namespace gdsl {
namespace assembly {

void scale::put(std::ostream& out) const { out << factor << *operand_; }

void scale::accept(operand_visitor& ov) const { ov.visit(this); }

bool scale::operator==(operand const& o) const {
  bool equals = false;
  operand_visitor v(true);
  v._([&](scale const* o) {
    equals = this->factor == o->factor && *this->operand_ == *o->operand_;
  });
  o.accept(v);
  return equals;
}

}  // namespace assembly
}  // namespace gdsl
